sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -4231,11 +4230,25 @@ static pi_result cleanupAfterEvent(pi_event Event) {
         }
 
         // It is possible that the fence was already noted as signalled and
-        // reset.  In that case the InUse flag will be false, and there is
-        // no need to query the fence's status or try to reset it.
+        // reset.  In that case the InUse flag will be false, and
+        // we shouldn't query it, synchronize on it, or try to reset it.
         if (it->second.InUse) {
+          // Workaround for VM_BIND mode.
+          // Make sure that the command-list doing memcpy is reset before
+          // non-USM host memory potentially involved in the memcpy is freed.
+          //
+          // NOTE: it is valid to wait for the fence here as long as we aren't
+          // doing batching on the involved command-list. Today memcpy goes by
+          // itself in a command list.
+          //
```

Comment:
Not sure if this NOTE: portion of the comment is accurate.  I think batching can still be done on memcpy, but if we do batching on memcpy, then first memcpy event in the command list that finishes will force a fence synchronization for the whole command list.  I also think it would be OK to batch, and just force a memcpy to be the last command put into a batch.

Reply:
> I also think it would be OK to batch, and just force a memcpy to be the last command put into a batch.

That's what I called no batching of memory copies, i.e. where same command-list  would contain multiple memory copies. Yes, memcpy must be the last command in the batch, i.e. close the batch, in order for this fence wait logic here to be correct.